base/gmp.jl

```diff
@@ -633,12 +633,17 @@ function gcdx(a::BigInt, b::BigInt)
 end
 
 sum(arr::AbstractArray{BigInt}) = foldl(MPZ.add!, arr; init=BigInt(0))
-# Note: a similar implementation for `prod` won't be efficient:
-# 1) the time complexity of the allocations is negligible compared to the multiplications
-# 2) assuming arr contains similarly sized BigInts, the multiplications are much more
-# performant when doing e.g. ((a1*a2)*(a3*a4))*(...) rather than a1*(a2*(a3*(...))),
-# which is exactly what the default implementation of `prod` does, via `mapreduce`
-# (which maybe could be slightly optimized for BigInt).
+
+function prod(arr::AbstractArray{BigInt})
+    # compute first the needed number of limbs for the result,
+    # to avoid re-allocations
+    nlimbs = sum(arr; init=0) do x
+        abs(x.size)
+    end
```

Comment:
I'd worry slightly this has exponential memory blowup. I realize that this is already the case for multiplication generally, but if most of `arr` was ones or other small numbers, this seems like it would vastly over-estimate the size of the final result. Then if `prod` is repeated, would it double the size of the allocation again? I notice that we don't estimate the size of `mul` normally.

 The old comment also claimed this performs the operations in a less efficient order? The benchmark you gave might be multiplying relatively few, small numbers (20) to see the advantage?

Reply:
I can experiment with making an estimate at the bit level instead of the limb level, this would limit the blowup.

> Then if prod is repeated, would it double the size of the allocation again? 
No, as this uses `x.size` which is exact, and can be less than the memory which was allocated (`x.alloc`). 

> The old comment also claimed this performs the operations in a less efficient order?

The old comment (by me), didn't take into account the possibility of pre-allocating, and underestimated the cost of allocations compared to that of the multiplication. I would tend to believe that the different order was more efficient (assuming equally sized integers) because there would be less intermediate allocations.